src/Storages/StorageMerge.cpp

```diff
@@ -616,22 +592,109 @@ SelectQueryInfo ReadFromMerge::getModifiedQueryInfo(const SelectQueryInfo & quer
+                        storage_metadata_snapshot->getColumns().getAll(), context);
+                    column_expr = setAlias(column_expr, column);
+
+                    auto type = sample_block.has(column) ? sample_block.getByName(column).type : column_description.type;
+
+                    aliases.push_back({ .name = column, .type = type, .expression = column_expr->clone() });
+                }
+                else
+                    column_expr = std::make_shared<ASTIdentifier>(column);
+
+                required_columns_expr_list->children.emplace_back(std::move(column_expr));
+            }
+
+            syntax_result = TreeRewriter(context).analyze(
+                required_columns_expr_list, storage_columns.getAllPhysical(), storage, storage->getStorageSnapshot(storage_metadata_snapshot, context));
+
+            auto alias_actions = ExpressionAnalyzer(required_columns_expr_list, syntax_result, context).getActionsDAG(true);
+
+            column_names_as_aliases = alias_actions->getRequiredColumns().getNames();
+
+            if (column_names_as_aliases.empty())
+                column_names_as_aliases.push_back(ExpressionActions::getSmallestColumn(storage_metadata_snapshot->getColumns().getAllPhysical()).name);
+        }
+    }
+    if (!column_names_as_aliases.empty())
+    {
+        real_column_names = column_names_as_aliases;
+    }
+}
+
+
 QueryPipelineBuilderPtr ReadFromMerge::createSources(
-    const StorageSnapshotPtr & storage_snapshot,
-    SelectQueryInfo & modified_query_info,
-    const QueryProcessingStage::Enum & processed_stage,
-    const UInt64 max_block_size,
+    QueryProcessingStage::Enum processed_stage,
+    UInt64 max_block_size,
     const Block & header,
-    const Aliases & aliases,
     const StorageWithLockAndName & storage_with_lock,
     Names real_column_names,
+    const Block & sample_block,
     ContextMutablePtr modified_context,
     size_t streams_num,
     bool concat_streams)
 {
     const auto & [database_name, storage, _, table_name] = storage_with_lock;
+    auto storage_metadata_snapshot = storage->getInMemoryMetadataPtr();
+    auto storage_snapshot = storage->getStorageSnapshot(storage_metadata_snapshot, context);
+    auto modified_query_info = getModifiedQueryInfo(query_info, context, storage_with_lock, storage_snapshot);
+
     auto & modified_select = modified_query_info.query->as<ASTSelectQuery &>();
 
+    std::unique_ptr<RowPolicyData> row_policy_data_ptr;
+
+    auto row_policy_filter_ptr = context->getRowPolicyFilter(
+        database_name,
+        table_name,
+        RowPolicyFilterType::SELECT_FILTER);
+    if (row_policy_filter_ptr)
+    {
+        row_policy_data_ptr = std::make_unique<RowPolicyData>(row_policy_filter_ptr, storage, context);
+        row_policy_data_ptr->extendNames(real_column_names);
+    }
+
+    Aliases aliases;
+    processAliases(real_column_names, storage_with_lock, aliases, sample_block, modified_context);
```

Comment:
Maybe it's better to rename `processAliases` to `getAliases()` and make it return `Aliases`?